benchmark/benchmain/main.go

```diff
@@ -365,178 +413,202 @@ func runBenchmark(caller func(int), startTimer func(), stopTimer func(uint64), b
+	compModes          []string        // Feature index 7
+	enableChannelz     []bool          // Feature index 8
+	enablePreloader    []bool          // Feature index 9
+}
+
+// generateFeatures generates all combinations of the provided feature options.
+// For example, let's say the user sets -workloads=all and
+// -maxConcurrentCalls=1,100, this would end up with the following
+// combinations:
+// [workloads: unary, maxConcurrentCalls=1]
+// [workloads: unary, maxConcurrentCalls=1]
+// [workloads: streaming, maxConcurrentCalls=100]
+// [workloads: streaming, maxConcurrentCalls=100]
+// [workloads: unconstrained, maxConcurrentCalls=1]
+// [workloads: unconstrained, maxConcurrentCalls=100]
+func (b *benchOpts) generateFeatures() []stats.Features {
+	// We use reflection here to get the number of features. Then we get the
+	// number of values for each feature. Each combination is represented using a
+	// slice (size=#features), where each item represents one feature. The value
+	// of each item is an index into a slice which contains all possible values
+	// for that feature. Based off of each combination, we create stats.Feature
+	// objects.
+	value := reflect.ValueOf(*b.features)
+	num := value.NumField()
+	featuresNum := make([]int, num)
+	for i := 0; i < num; i++ {
+		field := value.Field(i)
+		if field.Kind() != reflect.Slice {
+			log.Fatalf("Expect features to be slices, but found %v", field.Kind())
+		}
+		featuresNum[i] = field.Len()
+	}
+
+	var result []stats.Features
+	var curPos []int
+	initialPos := make([]int, num)
+	for !reflect.DeepEqual(initialPos, curPos) {
+		if curPos == nil {
+			curPos = make([]int, num)
+		}
+		result = append(result, stats.Features{
+			// These features stay the same for each iteration.
+			NetworkMode: b.networkMode,
+			UseBufConn:  b.useBufconn,
+			// These features can potentially change for each iteration.
+			EnableTrace:        b.features.enableTrace[curPos[0]],
+			Latency:            b.features.readLatencies[curPos[1]],
+			Kbps:               b.features.readKbps[curPos[2]],
+			Mtu:                b.features.readMTU[curPos[3]],
+			MaxConcurrentCalls: b.features.maxConcurrentCalls[curPos[4]],
+			ReqSizeBytes:       b.features.reqSizeBytes[curPos[5]],
+			RespSizeBytes:      b.features.respSizeBytes[curPos[6]],
+			ModeCompressor:     b.features.compModes[curPos[7]],
+			EnableChannelz:     b.features.enableChannelz[curPos[8]],
+			EnablePreloader:    b.features.enablePreloader[curPos[9]],
+		})
+		bm.AddOne(curPos, featuresNum)
+	}
+	return result
+}
+
+// sharedFeatures returns a bool slice which acts as a bitmask. Each item in
+// the slice represents a feature (which item corresponds to which feature can
+// be found in the comments for featureOpts). The bit is set to 1 if the
+// corresponding feature does not have multiple value, so is shared amongst all
+// benchmarks.
+func (b *benchOpts) sharedFeatures() []bool {
+	value := reflect.ValueOf(*b.features)
```

Comment:
Comments for generateFeatures applies here as well.

Reply:
One good thing with `sharedFeatures` and `generateFeatures` is that the `main` function is quite easy to read, unlike before.

Reply:
See my comments for `generateFeatures`